xdsl/parser.py

```diff
@@ -461,35 +463,36 @@ def parse_list(self,
         return res
 
     def parse_dictionary(self,
-                         parse_optional_one: Callable[[], T | None],
+                         parse_optional_key: Callable[[], K | None],
+                         parse_optional_value: Callable[[], V | None],
                          delimiter: str = ",",
-                         skip_white_space: bool = True) -> dict[T]:
+                         skip_white_space: bool = True) -> dict[K, V]:
         if skip_white_space:
             self.skip_white_space()
         assert (len(delimiter) <= 1)
-        res = dict[Any]()  # Pyright do not let us use `T` here
-        entry = self.parse_dict_entry(parse_optional_one)
-        if entry is not None:
-            res = res | entry
-        while self.parse_optional_char(delimiter) if len(
-                delimiter) == 1 else True:
-            entry = self.parse_dict_entry(parse_optional_one)
-            if entry is None:
-                return res
-            res = res | entry
+        res = self.parse_optional_dict_entry(parse_optional_key,
+                                             parse_optional_value)
+        if res is None:
+            return {}
+        while (self.parse_optional_char(delimiter)
+               if len(delimiter) == 1 else True):
+            entry = self.parse_optional_dict_entry(parse_optional_key,
+                                                   parse_optional_value)
+            if entry is not None:
+                res = res | entry
+
         return res
 
-    def parse_dict_entry(
+    def parse_optional_dict_entry(
         self,
-        parse_optional_one: Callable[[],
-                                     T | None]) -> dict[str, Attribute] | None:
-        # Limitation currently is that the key is a string
-        key = self.parse_str_literal()
-        if self.parse_optional_char("="):
-            value = parse_optional_one()
-            if value is not None:
-                return {key: value}
-        return None
+        parse_optional_key: Callable[[], K | None],
+        parse_optional_value: Callable[[], V | None],
+    ) -> dict[K, V] | None:
+        if not ((key := parse_optional_key()) is not None
```

Comment:
Once you parse the beginning, you should commit to parsing the whole thing.
Otherwise, we would return None when we already parsed a bit of it.

Reply:
I agree, I wanted not to change the logic in this PR, but am happy to do it if everyone's on board